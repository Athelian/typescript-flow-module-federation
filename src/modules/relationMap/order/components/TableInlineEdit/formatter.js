// @flow
import normalize from './normalize';

const array2Object = (inputArray: Array<any>): Object =>
  inputArray.reduce((result, key) => ({ ...result, [key]: true }), {});

export const formatOrders = ({
  orders,
  shipments,
}: {
  orders: Array<Object>,
  shipments: Array<Object>,
}) => {
  const { entities } = normalize({ orders, orderItems: [], batches: [], shipments });
  const order = {};
  const orderItem = {};
  const batch = {};
  const shipment = {};
  (Object.entries(entities.orders || {}): any).forEach(([orderId, orderData]) => {
    const currentOrder = orders.find(item => item.id === orderId);
    if (currentOrder) {
      order[orderId] = {
        data: currentOrder,
        relation: {
          orderItem: array2Object(orderData.orderItems),
          batch: array2Object(
            orderData.orderItems
              .map(orderItemId => entities.orderItems[orderItemId].batches)
              .reduce((result, currentBatches) => result.concat(currentBatches), [])
          ),
          shipment: array2Object(orderData.shipments),
        },
      };
      orderData.shipments.forEach(shipmentId => {
        const currentShipment = shipments.find(item => item.id === shipmentId);
        if (currentShipment) {
          if (!shipment[shipmentId]) {
            shipment[shipmentId] = {
              data: currentShipment,
              relation: {
                order: {
                  [orderId]: true,
                },
              },
            };
          } else {
            shipment[shipmentId].relation.order[orderId] = true;
          }
        }
      });

      orderData.orderItems.forEach(orderItemId => {
        const orderItemData = entities.orderItems[orderItemId];
        if (orderItemData) {
          orderItem[orderItemId] = {
            data: {
              ...orderItemData,
              batches: orderItemData.batches
                .map(batchId => entities.batches[batchId])
                .map(batchItem => ({
                  ...batchItem,
                  shipment: batchItem.shipment ? entities.shipments[batchItem.shipment] : null,
                })),
            },
            relation: {
              order: {
                [orderId]: true,
              },
              batch: array2Object(orderItemData.batches || []),
            },
          };
          (orderItemData.batches || []).forEach(batchId => {
            const batchData = entities.batches[batchId];
            batch[batchId] = {
              data: batchData,
              relation: {
                order: {
                  [orderId]: true,
                },
                orderItem: {
                  [orderItemId]: true,
                },
                shipment: {
                  ...(batchData.shipment ? { [batchData.shipment]: true } : {}),
                },
              },
            };
          });
        }
      });
    }
  });

  // Tracking all shipment which has no order/item/batch
  const shipmentNoRelation = {};
  shipments
    .filter(item => !shipment[item.id])
    .forEach(item => {
      shipment[item.id] = {
        data: item,
        relation: {
          batch: {},
          orderItem: {},
          order: {},
        },
      };
      shipmentNoRelation[item.id] = {
        data: item,
        relation: {
          batch: {},
          orderItem: {},
          order: {},
        },
      };
    });
  return {
    order,
    orderItem,
    batch,
    shipment,
    shipmentNoRelation,
  };
};

export default formatOrders;
