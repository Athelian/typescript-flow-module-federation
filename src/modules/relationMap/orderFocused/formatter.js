// @flow
import { pipe, filter, isEmpty } from 'ramda';

const prependArray = (arr: Array<any>) => (newArr: Array<any>) => [...arr, ...newArr];

const filterOutNewData = (newData: Array<Object>) => (data: Object) =>
  !newData ? true : !newData.some(d => d.id === data.id);

const createFilterData = (newData: Array<Object> = []) =>
  pipe(
    filter(filterOutNewData(newData || [])),
    prependArray(newData.map(d => Object.assign(d, { isNew: true })))
  );
const getIsNew = (item: Object) => item.isNew || false;

export const getIsCollapsed = (result: Object, order: Object) => {
  const { newOrderItem, newBatch } = result;
  if (newOrderItem[order.id] && !isEmpty(newOrderItem[order.id])) {
    return false;
  }
  if (!order.orderItems || (order.orderItems && isEmpty(order.orderItems))) {
    return true;
  }
  const hasBatchResult = (order.orderItems || []).some(({ id: orderItemId }) => {
    if (newBatch[orderItemId] && !isEmpty(newBatch[orderItemId])) {
      return true;
    }
    return false;
  });
  return !hasBatchResult;
};

export const formatNodes = (orders: Array<Object>, result: Object) => {
  const { order: newOrder = [], orderItem: newOrderItem = {}, batch: newBatch = {} } = result;
  const filterOrder = createFilterData(newOrder);
  const nodes: Array<Object> = filterOrder(orders).map(order => {
    const filterOrderItem = createFilterData(newOrderItem[order.id]);
    const orderItems = filterOrderItem(order.orderItems || []).map(orderItem => {
      const filterBatch = createFilterData(newBatch[orderItem.id]);
      const batches = filterBatch(orderItem.batches || []).map(batch => ({
        id: batch.id,
        isNew: getIsNew(batch),
      }));
      return {
        id: orderItem.id,
        isNew: getIsNew(orderItem),
        batches,
      };
    });
    return {
      id: order.id,
      isNew: getIsNew(order),
      isCollapsed: getIsCollapsed({ newOrderItem, newBatch }, order),
      orderItems,
    };
  });
  return nodes;
};

export default null;
