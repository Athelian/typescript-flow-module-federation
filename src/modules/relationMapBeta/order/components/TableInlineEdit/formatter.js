// @flow
import {
  initOrderObj,
  initOrderItemObj,
  initBatchObj,
  initShipmentObj,
} from 'modules/relationMap/util';

function* iterateOrders(orders: Array<Object>) {
  for (let orderIndex = 0; orderIndex < orders.length; orderIndex += 1) {
    const order = orders[orderIndex];
    yield { type: 'ORDER', index: orderIndex, data: order };
    const { orderItems = [], shipments = [] } = order;
    for (let shipmentIndex = 0; shipmentIndex < shipments.length; shipmentIndex += 1) {
      const shipment = shipments[shipmentIndex];
      yield { type: 'SHIPMENT', index: shipmentIndex, data: shipment };
    }
    for (let itemIndex = 0; itemIndex < orderItems.length; itemIndex += 1) {
      const orderItem = orderItems[itemIndex];
      yield { type: 'ORDER_ITEM', index: itemIndex, data: orderItem };
      const { batches = [] } = orderItem;
      for (let batchIndex = 0; batchIndex < batches.length; batchIndex += 1) {
        const batch = batches[batchIndex];
        yield { type: 'BATCH', index: batchIndex, data: batch };
      }
    }
  }
}

const getBatchedQuantity = (batch: Object) => {
  const { quantity, batchAdjustments = [] } = batch;
  return batchAdjustments.reduce((total, adjustment) => total + adjustment.quantity, quantity || 0);
};

const getBatchRelation = (batch: Object, info: Object) => {
  const { orderId, orderItemId, index } = info;
  return {
    ...batch,
    rootId: orderId,
    parentId: orderItemId,
    index,
  };
};

const summary = (orders: Array<Object>) => {
  const sumOrders = orders.length;
  let sumOrderItems = 0;
  let sumBatches = 0;
  let sumShipments = 0;

  const getSummary = () => ({
    sumOrders,
    sumOrderItems,
    sumBatches,
    sumShipments,
  });

  const findSummary = (entity: Object) => {
    if (entity.type === 'ORDER') {
      const { orderItems = [], shipments = [] } = entity.data;
      sumOrderItems += orderItems.length;
      sumShipments += shipments.length;
    }
    if (entity.type === 'ORDER_ITEM') {
      const { batches = [] } = entity.data;
      sumBatches += batches.length;
    }
    return getSummary();
  };
  return {
    findSummary,
    getSummary,
  };
};

const createOrderObj = () => {
  const orderObj = {};
  let orderId = '';
  let orderItemId = '';
  let batchId = '';
  const getOrderObj = () => ({ order: orderObj });
  const formatOrderObj = (entity: Object) => {
    const { type, data, index } = entity;
    const { id } = data;
    if (type === 'ORDER') {
      if (!orderObj[id]) {
        orderObj[id] = initOrderObj(data);
      }
      orderId = id;
    }
    if (type === 'ORDER_ITEM') {
      const { batches = [] } = data;
      const { relation: orderRelation, data: orderData } = orderObj[orderId];
      orderItemId = id;
      orderRelation.orderItem[orderItemId] = {
        ...data,
        parentId: orderId,
        index: orderItemId,
      };
      orderData.totalBatch += batches.length;
      orderData.orderedQuantity += data.quantity || 0;
    }
    if (type === 'BATCH') {
      const { shipment } = data;
      const { relation: orderRelation, data: orderData } = orderObj[orderId];
      const batchedQuantity = getBatchedQuantity(data);
      batchId = id;
      orderData.batchedQuantity += batchedQuantity;
      orderRelation.batch[batchId] = getBatchRelation(data, {
        orderId,
        orderItemId,
        index,
      });
      if (shipment) {
        orderData.shippedQuantity += batchedQuantity;
        orderRelation.shipment[shipment.id] = true;
      }
    }
  };
  return {
    formatOrderObj,
    getOrderObj,
  };
};

const createOrderItemObj = () => {
  let orderId = '';
  let order = {};
  let orderItemId = '';
  const orderItemObj = {};
  const relatedShipment = {};
  const getOrderItemObj = () => ({ orderItem: orderItemObj });
  const formatOrderItemObj = entity => {
    const { type, data, index } = entity;
    const { id } = data;
    if (type === 'ORDER') {
      orderId = id;
      order = data;
      const { shipments } = order;
      shipments.forEach(shipment => {
        const batches = shipment.batches || [];
        batches.forEach(batch => {
          relatedShipment[batch.id] = shipment;
        });
      });
    }
    if (type === 'ORDER_ITEM') {
      orderItemId = id;
      if (!orderItemObj[id]) {
        orderItemObj[id] = initOrderItemObj(data, order);
        orderItemObj[id].data.batches.forEach((batch, batchIndex) => {
          if (batch) {
            const batchData = orderItemObj[id].data.batches[batchIndex];
            batchData.shipment = relatedShipment[batch.id] || {};
          }
        });
      }
    }
    if (type === 'BATCH') {
      const { shipment } = data;
      const batchedQuantity = getBatchedQuantity(data);
      const { relation: orderItemRelation, data: orderItemData } = orderItemObj[orderItemId];
      orderItemRelation.batch[id] = getBatchRelation(data, {
        orderId,
        orderItemId,
        index,
      });
      orderItemData.batchedQuantity += batchedQuantity;
      if (shipment) {
        orderItemData.shippedQuantity += batchedQuantity;
        orderItemRelation.shipment[shipment.id] = true;
      }
    }
  };
  return {
    formatOrderItemObj,
    getOrderItemObj,
  };
};

const createBatchObj = () => {
  let order = {};
  let orderItem = {};
  let orderId = '';
  const batchObj = {};
  const getBatchObj = () => ({ batch: batchObj });
  const formatBatchObj = entity => {
    const { type, data } = entity;
    const { id } = data;
    if (type === 'ORDER') {
      orderId = id;
      order = data;
    }
    if (type === 'ORDER_ITEM') {
      orderItem = { ...data, order };
    }
    if (type === 'BATCH') {
      if (!batchObj[id]) {
        batchObj[id] = initBatchObj(data, orderId, orderItem);
      }
      const { data: batchData } = batchObj[id];
      batchData.batchedQuantity = getBatchedQuantity(data);
    }
  };
  return {
    getBatchObj,
    formatBatchObj,
  };
};

const createShipmentObj = () => {
  const shipmentObj = {};
  let orderId = '';
  let orderItemId = '';
  const getShipmentObj = () => ({ shipment: shipmentObj });
  const formatShipmentObj = entity => {
    const { type, data } = entity;
    const { id } = data;
    if (type === 'ORDER') {
      orderId = id;
    }
    if (type === 'ORDER_ITEM') {
      orderItemId = id;
    }
    if (type === 'SHIPMENT') {
      if (!shipmentObj[id]) {
        shipmentObj[id] = initShipmentObj(data);
      }
      const { relation: shipmentRelation } = shipmentObj[id];
      shipmentRelation.order[orderId] = true;
    }
    if (type === 'BATCH') {
      const { shipment } = data;
      if (shipment) {
        const { id: shipmentId } = shipment;
        if (shipment && !shipmentObj[shipmentId]) {
          shipmentObj[shipmentId] = initShipmentObj(shipment);
        }
        const { relation: shipmentRelation } = shipmentObj[shipmentId];
        shipmentRelation.order[orderId] = true;
        shipmentRelation.orderItem[orderItemId] = true;
        shipmentRelation.batch[id] = true;
      }
    }
  };
  return {
    getShipmentObj,
    formatShipmentObj,
  };
};

export const formatOrders = (orders: Array<Object>) => {
  const orderGenerator = iterateOrders(orders);
  const summaryData = summary(orders);
  const orderData = createOrderObj();
  const itemData = createOrderItemObj();
  const batchData = createBatchObj();
  const shipmentData = createShipmentObj();
  let result = orderGenerator.next();
  while (!result.done) {
    summaryData.findSummary(result.value);
    orderData.formatOrderObj(result.value);
    itemData.formatOrderItemObj(result.value);
    batchData.formatBatchObj(result.value);
    shipmentData.formatShipmentObj(result.value);
    result = orderGenerator.next();
  }
  return {
    ...summaryData.getSummary(),
    ...orderData.getOrderObj(),
    ...itemData.getOrderItemObj(),
    ...batchData.getBatchObj(),
    ...shipmentData.getShipmentObj(),
  };
};

export default formatOrders;
