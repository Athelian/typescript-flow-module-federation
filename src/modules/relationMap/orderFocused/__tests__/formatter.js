import { formatNodes, getIsCollapsed, formatOrders } from '../formatter';
import ordersData from './data.json';

describe('_formatNodes', () => {
  it('format order', () => {
    const data = [
      { id: 1, orderItems: [] },
      { id: 2, orderItems: [] },
      { id: 3, orderItems: [] },
      { id: 4, orderItems: [] },
    ];
    const result = { order: [{ id: 2 }, { id: 4 }] };
    const expected = formatNodes(data, result);
    expect(expected).toMatchSnapshot();
  });

  it('format newOrderItem', () => {
    const data = [
      {
        id: 1,
        orderItems: [
          { id: 1, batches: [] },
          { id: 2, batches: [] },
          { id: 3, batches: [] },
          { id: 4, batches: [] },
        ],
      },
    ];
    const result = {
      orderItem: {
        '1': [{ id: 4 }],
      },
    };
    const expected = formatNodes(data, result);
    expect(expected).toMatchSnapshot();
  });

  it('format batches', () => {
    const data = [
      {
        id: 1,
        orderItems: [
          {
            id: 1,
            batches: [{ id: 1 }, { id: 11 }, { id: 12 }, { id: 13 }, { id: 14 }],
          },
        ],
      },
    ];
    const result = {
      batch: {
        1: [{ id: 12 }, { id: 14 }],
      },
    };
    const expected = formatNodes(data, result);
    expect(expected).toMatchSnapshot();
  });

  it('format mess data', () => {
    const data = [
      {
        id: 1,
        orderItems: [
          {
            id: 1,
            batches: [{ id: 1 }, { id: 11 }, { id: 12 }, { id: 13 }, { id: 14 }],
          },
        ],
      },
      {
        id: 2,
        orderItems: [{ id: 2, batches: null }, { id: 3, batches: null }, { id: 4, batches: null }],
      },
    ];
    const result = {
      orderItem: {
        2: [{ id: 4 }],
      },
      batch: {
        1: [{ id: 12 }, { id: 14 }],
      },
    };
    const expected = formatNodes(data, result);
    expect(expected).toMatchSnapshot();
  });
});

describe('getIsCollapsed', () => {
  it('no orderItem', () => {
    const order = {
      id: 1,
      orderItems: [],
    };
    const newResult = {
      newOrderItem: {},
      newBatch: {},
    };
    const result = getIsCollapsed(newResult, order);
    expect(result).toEqual(true);
  });

  it('has orderItem', () => {
    const order = {
      id: 1,
      orderItems: [],
    };
    const newResult = {
      newOrderItem: {
        '1': ['1', '2'],
      },
      newBatch: {},
    };
    const result = getIsCollapsed(newResult, order);
    expect(result).toEqual(false);
  });

  it('has new orderItem', () => {
    const order = {
      id: 1,
      orderItems: [
        {
          id: '1',
        },
        {
          id: '2',
        },
      ],
    };
    const newResult = {
      newOrderItem: {},
      newBatch: {
        '1': [{ id: 1 }],
      },
    };
    const result = getIsCollapsed(newResult, order);
    expect(result).toEqual(false);
  });

  it('has new orderItem with batch', () => {
    const order = {
      id: '1',
      orderItems: [
        {
          id: '1',
          batches: [{ id: '1' }],
        },
        {
          id: '2',
        },
      ],
    };
    const newResult = {
      newOrderItem: {
        '1': [{ id: '1' }],
      },
      newBatch: {},
    };
    const result = getIsCollapsed(newResult, order);
    expect(result).toEqual(false);
  });
  it('has not new orderItem', () => {
    const order = {
      id: 1,
      orderItems: [
        {
          id: 1,
        },
        {
          id: 2,
        },
      ],
    };
    const newResult = {
      newOrderItem: {},
      newBatch: {},
    };
    const result = getIsCollapsed(newResult, order);
    expect(result).toEqual(true);
  });
});

describe('formatOrders', () => {
  it.only('test format order data', () => {
    const result = formatOrders(ordersData);
    expect(result).toMatchSnapshot();
  });
});
