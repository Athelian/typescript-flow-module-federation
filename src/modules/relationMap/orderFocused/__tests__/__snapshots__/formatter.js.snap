// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_formatNodes format batches 1`] = `
Array [
  Object {
    "id": 1,
    "orderItems": Array [
      Object {
        "batches": Array [
          Object {
            "id": 12,
          },
          Object {
            "id": 14,
          },
          Object {
            "id": 1,
          },
          Object {
            "id": 11,
          },
          Object {
            "id": 13,
          },
        ],
        "id": 1,
      },
    ],
  },
]
`;

exports[`_formatNodes format mess data 1`] = `
Array [
  Object {
    "id": 1,
    "orderItems": Array [
      Object {
        "batches": Array [
          Object {
            "id": 12,
          },
          Object {
            "id": 14,
          },
          Object {
            "id": 1,
          },
          Object {
            "id": 11,
          },
          Object {
            "id": 13,
          },
        ],
        "id": 1,
      },
    ],
  },
  Object {
    "id": 2,
    "orderItems": Array [
      Object {
        "batches": Array [],
        "id": 4,
      },
      Object {
        "batches": Array [],
        "id": 2,
      },
      Object {
        "batches": Array [],
        "id": 3,
      },
    ],
  },
]
`;

exports[`_formatNodes format newOrderItem 1`] = `
Array [
  Object {
    "id": 1,
    "orderItems": Array [
      Object {
        "batches": Array [],
        "id": 4,
      },
      Object {
        "batches": Array [],
        "id": 1,
      },
      Object {
        "batches": Array [],
        "id": 2,
      },
      Object {
        "batches": Array [],
        "id": 3,
      },
    ],
  },
]
`;

exports[`_formatNodes format order 1`] = `
Array [
  Object {
    "id": 2,
    "orderItems": Array [],
  },
  Object {
    "id": 4,
    "orderItems": Array [],
  },
  Object {
    "id": 1,
    "orderItems": Array [],
  },
  Object {
    "id": 3,
    "orderItems": Array [],
  },
]
`;
